tools/handshake-runner/Simulation.cpp

```diff
@@ -0,0 +1,722 @@
+  LLVM_DEBUG(dbgs() << "  " << head << ":  " << op << " = ";
+             value.print(dbgs()); dbgs() << " ("
+                                         << "float"
+                                         << ") @" << time << "\n");
+}
+
+void debugArg(const std::string &head, mlir::Value op, const Any &value,
+              double time) {
+  if (any_isa<APInt>(value)) {
+    debugArg(head, op, any_cast<APInt>(value), time);
+  } else if (any_isa<APFloat>(value)) {
+    debugArg(head, op, any_cast<APFloat>(value), time);
+  } else if (any_isa<unsigned>(value)) {
+    // Represents an allocated buffer.
+    LLVM_DEBUG(dbgs() << "  " << head << ":  " << op << " = Buffer "
+                      << any_cast<unsigned>(value) << "\n");
+  } else {
+    llvm_unreachable("unknown type");
+  }
+}
+
+Any readValueWithType(mlir::Type type, std::string in) {
+  std::stringstream arg(in);
+  if (type.isIndex()) {
+    int64_t x;
+    arg >> x;
+    int64_t width = INDEX_WIDTH;
+    APInt aparg(width, x);
+    return aparg;
+  } else if (type.isa<mlir::IntegerType>()) {
+    int64_t x;
+    arg >> x;
+    int64_t width = type.getIntOrFloatBitWidth();
+    APInt aparg(width, x);
+    return aparg;
+    // } else if (type.isF16()) {
+    //   half x;
+    //   arg >> x;
+    //   APFloat aparg(x);
+    //   return aparg;
+  } else if (type.isF32()) {
+    float x;
+    arg >> x;
+    APFloat aparg(x);
+    return aparg;
+  } else if (type.isF64()) {
+    double x;
+    arg >> x;
+    APFloat aparg(x);
+    return aparg;
+  } else {
+    assert(0 && "unknown argument type!\n");
+  }
+}
+
+std::string printAnyValueWithType(mlir::Type type, Any &value) {
+  std::stringstream out;
+  if (type.isa<mlir::IntegerType>() || type.isa<mlir::IndexType>()) {
+    out << any_cast<APInt>(value).getSExtValue();
+    return out.str();
+  } else if (type.isa<mlir::FloatType>()) {
+    out << any_cast<APFloat>(value).convertToDouble();
+    return out.str();
+  } else if (type.isa<mlir::NoneType>()) {
+    return "none";
+  } else {
+    llvm_unreachable("Unknown result type!");
+  }
+}
+
+void scheduleIfNeeded(std::list<mlir::Operation *> &readyList,
+                      llvm::DenseMap<mlir::Value, Any> &valueMap,
+                      mlir::Operation *op) {
+  if (std::find(readyList.begin(), readyList.end(), op) == readyList.end()) {
+    readyList.push_back(op);
+  }
+}
+void scheduleUses(std::list<mlir::Operation *> &readyList,
+                  llvm::DenseMap<mlir::Value, Any> &valueMap,
+                  mlir::Value value) {
+  for (auto &use : value.getUses()) {
+    scheduleIfNeeded(readyList, valueMap, use.getOwner());
+  }
+}
+
+bool executeStdOp(mlir::Operation &op, std::vector<Any> &inValues,
+                  std::vector<Any> &outValues) {
+  if (auto stdOp = dyn_cast<mlir::ConstantIndexOp>(op))
+    executeOp(stdOp, inValues, outValues);
+  else if (auto stdOp = dyn_cast<mlir::ConstantIntOp>(op))
+    executeOp(stdOp, inValues, outValues);
```

Comment:
Would it make sense to use the `dispatchStdExprVisitor()` of `inclue/circt/Dialect/Handshake/visitor.h` to refactor this implementation? A similar simplification is also used in the HandshakeToFIRRTL conversion pass.

Reply:
Thanks for your advice. Yes, I agree this can be optimized. This patch is mainly to rewrite the handshake part, so I suggest to leave the standard part as it is for now (I only did some renaming, so clang-tidy does not complain).

Reply:
Make sense to me. Let's first land this patch.